src/__private_api.rs

```diff
@@ -52,57 +52,43 @@ impl Log for GlobalLogger {
     }
 }
 
-// Split from `log` to reduce generics and code size
-fn log_impl<L: Log>(
-    logger: L,
-    args: Arguments,
-    level: Level,
-    &(target, module_path, loc): &(&str, &'static str, &'static Location),
-    kvs: Option<&[(&str, Value)]>,
-) {
-    #[cfg(not(feature = "kv"))]
-    if kvs.is_some() {
-        panic!("key-value support is experimental and must be enabled using the `kv` feature")
+pub trait LogExt: Log {
+    fn do_enabled(&self, level: Level, target: &str) -> bool {
+        self.enabled(&Metadata::builder().level(level).target(target).build())
     }
 
-    let mut builder = Record::builder();
-
-    builder
-        .args(args)
-        .level(level)
-        .target(target)
-        .module_path_static(Some(module_path))
-        .file_static(Some(loc.file()))
-        .line(Some(loc.line()));
-
-    #[cfg(feature = "kv")]
-    builder.key_values(&kvs);
-
-    logger.log(&builder.build());
-}
-
-pub fn log<'a, K, L>(
-    logger: L,
-    args: Arguments,
-    level: Level,
-    target_module_path_and_loc: &(&str, &'static str, &'static Location),
-    kvs: K,
-) where
-    K: KVs<'a>,
-    L: Log,
-{
-    log_impl(
-        logger,
-        args,
-        level,
-        target_module_path_and_loc,
-        kvs.into_kvs(),
-    )
+    fn do_log<'a>(
+        &self,
```

Comment:
This forces caller to pass a non-ZST reference, which causes the ZST optimization to lose effect.

Reply:
I can see that the final `fn log(&self, ..)` and `fn enabled(&self, ..)` method take a reference also.

So the optimization happens only in the caller-side, as you described in https://github.com/rust-lang/log/pull/576#discussion_r2009369123?

Reply:
Yes. You need to make sure that the arguments passed to the outermost function are ZSTs.

Reply:
OK. Then I'd close this PR as not work.

Although, I wonder if `write!` use `$dst.write_fmt` that accepts a reference already, is this optimization significant? Or a net win?